native/tests/unit-test/src/base_has_member_function_test.cpp

```diff
@@ -0,0 +1,159 @@
+ subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#include "base/HasMemberFunction.h"
+#include "core/scene-graph/Node.h"
+#include "gtest/gtest.h"
+
+using namespace cc;
+
+namespace {
+
+class BaseWithoutGetScriptObject {};
+
+class BaseWithGetScriptObjectButDifferentSignature {
+public:
+    int getScriptObject(bool) const { return false; }
+    float getScriptObject(int) const { return 1.F; }
+    inline Node* getScriptObject() { return _node; }
+    inline const Node* getScriptObject() const { return _node; }
+
+private:
+    Node* _node{nullptr};
+};
+
+class BaseWithCorrectGetScriptObjectSignature {
+public:
+    se::Object* getScriptObject() const { return nullptr; }
+    se::Object* getScriptObject() { return nullptr; }
+};
+
+static const unsigned long fakeAddr = 100;
+
+class Base0 {
+public:
+    Base0() {
+        _seObj = reinterpret_cast<se::Object*>(reinterpret_cast<uintptr_t>(&fakeAddr));
+    }
+
+    se::Object* getScriptObject() const { return _seObj; }
+    se::Object* getScriptObject() { return _seObj; }
+
+private:
+    se::Object* _seObj{nullptr};
+};
+
+class Base1 {
+public:
+};
+
+class Sub0 : public Base0 {
+};
+
+class Sub1 : public Base0, public Base1 {
+};
+
+class Sub2 : public Base1, public Base0 {
+};
+
+class BaseVirtual {
+public:
+    virtual ~BaseVirtual() = default;
+    virtual se::Object* getScriptObject() const = 0;
+    virtual se::Object* getScriptObject() = 0;
+};
+
+class BaseVirtualOverride : public BaseVirtual {
+public:
+    se::Object* getScriptObject() const override { return nullptr; }
+    se::Object* getScriptObject() override { return nullptr; }
+};
+
+class SubVirtualOverride : public BaseVirtualOverride {
+public:
+    SubVirtualOverride() {
+        _seObj = reinterpret_cast<se::Object*>(reinterpret_cast<uintptr_t>(&fakeAddr));
+    }
+    se::Object* getScriptObject() const override { return _seObj; }
+    se::Object* getScriptObject() override { return _seObj; }
+    int getScriptObject(int) { return 100; }
+    float getScriptObject(float) { return 200.F; }
+
+private:
+    se::Object* _seObj{nullptr};
+};
+
+} // namespace
+
+TEST(HasMemberFunction, test01) {
+    static_assert(!cc::has_getScriptObject<BaseWithoutGetScriptObject, se::Object*()>::value);
```

Comment:
I think it is better if the usage is
```c++
cc::hasMemberFunction<BaseWithoutGetScriptObject, getScriptObject, se::Object*()>::value>
```

Reply:
Can't change it to this.
You could refer to the implementation of `CC_DEFINE_HAS_MEMBER_FUNC` macro.
```c++
#define CC_DEFINE_HAS_MEMBER_FUNC(memFunc)                                    \
    template <typename, typename T>                                           \
    struct has_##memFunc {                                                    \
        static_assert(                                                        \
            std::integral_constant<T, false>::value,                          \
            "Second template parameter needs to be of function type.");       \
    };                                                                        \
                                                                              \
    template <typename C, typename Ret, typename... Args>                     \
    struct has_##memFunc<C, Ret(Args...)> {                                   \
    private:                                                                  \
        template <typename T>                                                 \
        static constexpr auto check(T*)                                       \
            -> typename std::is_same<                                         \
                decltype(std::declval<T>().memFunc(std::declval<Args>()...)), \
                Ret>::type;                                                   \
                                                                              \
        template <typename>                                                   \
        static constexpr std::false_type check(...);                          \
        typedef decltype(check<C>(0)) type;                                   \
                                                                              \
    public:                                                                   \
        static constexpr bool value = type::value;                            \
    };
```

`memFunc` should be defined since `check` function needs to invoke it. `memFunc` could not be a template type that passed to `cc::hasMemberFunction<BaseWithoutGetScriptObject, getScriptObject, se::Object*()>::value>`.
`cc::hasMemberFunction<BaseWithoutGetScriptObject, getScriptObject, se::Object*()>::value>` treats getScriptObject as a type which will not work.